package mesos;

option java_outer_classname = "Protos";


message FrameworkID {
  required string value = 1;
}


message OfferID {
  required string value = 1;
}


message SlaveID {
  required string value = 1;
}


message TaskID {
  required string value = 1;
}


message ExecutorID {
  required string value = 1;
}


message Param {
  required string key = 1;
  required string value = 2;
}


// TODO(benh): Eliminate this once some notion of a protocol buffer
// ResourceOffer object makes it's way to the Mesos interface.
message Params {
  repeated Param param = 1;
}


// TODO(benh): Add support for attributes.
// message Attribute {
//   required string name = 1;
//   required string value = 2;
// }


// TODO(benh): Add better support for resources.
message Resource {
  enum Type {
    SCALAR = 0;
    RANGES = 1;
    SET = 2;
  }

  message Scalar {
    required double value = 1;
  }

  message Range {
    required uint64 begin = 1;
    required uint64 end = 2;
  }

  message Ranges {
    repeated Range range = 1;
  }

  message Set {
    repeated string item = 1; 
  }

  required string name = 1;
  required Type type = 2;
  optional Scalar scalar = 3;
  optional Ranges ranges = 4;
  optional Set set = 5;
}


message ExecutorArgs {
  required FrameworkID framework_id = 1;
  required string name = 2;
  required SlaveID slave_id = 3;
  required string hostname = 4;
  optional bytes data = 5;
};


message StatusUpdate {
  required TaskID task_id = 1;
  required SlaveID slave_id = 2;
  required TaskState state = 3;
  optional bytes data = 4;
}


message TaskDescription {
  required string name = 1;
  required TaskID task_id = 2;
  required SlaveID slave_id = 3;
  repeated Resource resources = 4;
  optional ExecutorInfo executor = 5;
  optional bytes data = 6;
}


enum TaskState {
  TASK_STARTING = 0;
  TASK_RUNNING = 1;
  TASK_FINISHED = 2;
  TASK_FAILED = 3;
  TASK_KILLED = 4;
  TASK_LOST = 5;
}


message TaskStatus {
  required TaskID task_id = 1;
  required SlaveID slave_id = 2;
  required TaskState state = 3;
  optional bytes data = 4;
}


message ExecutorInfo {
  required ExecutorID executor_id = 1;
  required string uri = 2;
  optional Params params = 3;
  optional bytes data = 4;
}


message FrameworkInfo {
  required string user = 1;
  required string name = 2;
  required ExecutorInfo executor = 3;
}


message SlaveInfo {
  required string hostname = 1;
  required string public_hostname = 2;
  repeated Resource resources = 3;
//   repeated Attribute attribute = ;
}


message FrameworkMessage {
  required SlaveID slave_id = 1;
  required ExecutorID executor_id = 2;
  optional bytes data = 3;
}


message SlaveOffer {
  required SlaveID slave_id = 1;
  required string hostname = 2;
  repeated Resource resources = 3;
}
